libnetdata/dictionary/dictionary.c

```diff
@@ -533,636 +751,816 @@ static inline int hashtable_delete_unsafe(DICTIONARY *dict, const char *name, si
+    if(unlikely(!name || !*name)) {
+        internal_error(
+            true,
+            "DICTIONARY: attempted to %s() without a name on a dictionary created from %s() %zu@%s.",
+            __FUNCTION__,
+            dict->creation_function,
+            dict->creation_line,
+            dict->creation_file);
+        return NULL;
+    }
+
+    if(unlikely(is_dictionary_destroyed(dict))) {
+        internal_error(true, "DICTIONARY: attempted to dictionary_set() on a destroyed dictionary");
+        return NULL;
+    }
+
+    if(name_len == -1)
+        name_len = (ssize_t)strlen(name) + 1; // we need the terminating null too
+
+    debug(D_DICTIONARY, "SET dictionary entry with name '%s'.", name);
 
-    NAME_VALUE *nv;
+    // DISCUSSION:
+    // Is it better to gain a read-lock and do a hashtable_get_unsafe()
+    // before we write lock to do hashtable_insert_unsafe()?
+    //
+    // Probably this depends on the use case.
+    // For statsd for example that does dictionary_set() to update received values,
+    // it could be beneficial to do a get() before we insert().
+    //
+    // But the caller has the option to do this on his/her own.
+    // So, let's do the fastest here and let the caller decide the flow of calls.
 
-    debug(D_DICTIONARY, "Destroying dictionary.");
+    dictionary_index_lock_wrlock(dict);
 
-    long referenced_items = 0;
-    size_t retries = 0;
+    bool added_or_updated = false;
+    size_t spins = 0;
+    DICTIONARY_ITEM *item = NULL;
     do {
-        referenced_items = __atomic_load_n(&dict->referenced_items, __ATOMIC_SEQ_CST);
-        if (referenced_items) {
-            dictionary_lock(dict, DICTIONARY_LOCK_WRITE);
-
-            // there are referenced items
-            // delete all items individually, so that only the referenced will remain
-            NAME_VALUE *nv_next;
-            for (nv = dict->first_item; nv; nv = nv_next) {
-                nv_next = nv->next;
-                size_t refcount = DICTIONARY_NAME_VALUE_REFCOUNT_GET(nv);
-                if (!refcount && !(nv->flags & NAME_VALUE_FLAG_DELETED))
-                    dictionary_del_unsafe(dict, namevalue_get_name(nv));
+        DICTIONARY_ITEM **item_pptr = (DICTIONARY_ITEM **)hashtable_insert_unsafe(dict, name, name_len);
+        if (likely(*item_pptr == 0)) {
+            // a new item added to the index
+
+            // create the dictionary item
+            item = *item_pptr = item_create_with_hooks(dict, name, name_len, value, value_len, constructor_data);
+
+            // call the hashtable react
+            hashtable_inserted_item_unsafe(dict, item);
+
+            // unlock the index lock, before we add it to the linked list
+            // DONT DO IT THE OTHER WAY AROUND - DO NOT CROSS THE LOCKS!
+            dictionary_index_lock_unlock(dict);
+
+            item_linked_list_add(dict, item);
```

Comment:
And this is the last for this PR:

```sh
*** CID 357073:    (RESOURCE_LEAK)
/libnetdata/dictionary/dictionary.c: 1123 in item_add_or_reset_value_and_acquire()
1117                 item_linked_list_add(dict, item);
1118                 added_or_updated = true;
1119             }
1120             else {
1121                 if (!item_check_and_acquire(dict, *item_pptr)) {
1122                     spins++;
>>>     CID 357073:    (RESOURCE_LEAK)
>>>     Variable "item_pptr" going out of scope leaks the storage it points to.
1123                     continue;
1124                 }
1125     
1126                 // the item is already in the index
1127                 // so, either we will return the old one
1128                 // or overwrite the value, depending on dictionary flags
/libnetdata/dictionary/dictionary.c: 1160 in item_add_or_reset_value_and_acquire()
1154                     // we did really nothing!
1155                     ;
1156                 }
1157     
1158                 dictionary_index_lock_unlock(dict);
1159             }
>>>     CID 357073:    (RESOURCE_LEAK)
>>>     Variable "item_pptr" going out of scope leaks the storage it points to.
1160         } while(!item);
1161     
1162     
1163         if(unlikely(spins > 0 && dict->stats))
1164             DICTIONARY_STATS_INSERT_SPINS_PLUS(dict, spins);
```

Reply:
This is how Judy works. It gives you a pointer to its own storage which you can modify its contents to put whatever you like.